src/app/src/main/kotlin/org/vaccineimpact/api/app/controllers/CoverageController.kt

```diff
@@ -2,24 +2,33 @@ package org.vaccineimpact.api.app.controllers
 
 import org.vaccineimpact.api.app.app_start.Controller
 import org.vaccineimpact.api.app.context.ActionContext
+import org.vaccineimpact.api.app.context.RequestBodySource
 import org.vaccineimpact.api.app.controllers.helpers.ResponsibilityPath
 import org.vaccineimpact.api.app.logic.CoverageLogic
 import org.vaccineimpact.api.app.logic.RepositoriesCoverageLogic
 import org.vaccineimpact.api.app.repositories.Repositories
+import org.vaccineimpact.api.app.requests.csvData
+import org.vaccineimpact.api.app.requests.PostDataHelper
 import org.vaccineimpact.api.app.security.checkIsAllowedToSeeTouchstone
-import org.vaccineimpact.api.models.CoverageRow
-import org.vaccineimpact.api.models.ScenarioTouchstoneAndCoverageSets
+import org.vaccineimpact.api.models.*
 import org.vaccineimpact.api.serialization.SplitData
 import org.vaccineimpact.api.serialization.StreamSerializable
 
 class CoverageController(
         actionContext: ActionContext,
-        private val coverageLogic: CoverageLogic
+        private val coverageLogic: CoverageLogic,
+        private val postDataHelper: PostDataHelper = PostDataHelper()
 ) : Controller(actionContext)
 {
     constructor(context: ActionContext, repositories: Repositories)
             : this(context, RepositoriesCoverageLogic(repositories))
 
+    fun getCoverageDataFromCSV(): Sequence<CoverageIngestionRow>
+    {
+        val source = RequestBodySource(context)
+        return postDataHelper.csvData(from = source)
```

Comment:
Oh that's clever, does it work out the class to deserialize to from the method's return type?

Reply:
I know, clever compiler right!